tests/test_generate_data.py

```diff
@@ -3,12 +3,464 @@
+
+    def test_generate(self):
+        with patch("logging.Logger.info") as mocked_logger:
+            generate_data(
+                mocked_logger,
+                model_family="merlinite",
+                model_name="models/merlinite-7b-lab-Q4_K_M.gguf",
+                num_instructions_to_generate=10,
+                taxonomy=self.test_taxonomy.root,
+                taxonomy_base=TEST_TAXONOMY_BASE,
+                output_dir=self.tmp_path,
+                yaml_rules=TEST_CUSTOM_YAML_RULES,
+                chunk_word_count=1000,
+                server_ctx_size=4096,
+                client=MagicMock(),
+                pipeline="simple",
+            )
+
+            for name in ["test_*.jsonl", "train_*.jsonl", "messages_*.jsonl"]:
+                matches = glob.glob(os.path.join(self.tmp_path, name))
+                assert len(matches) == 1
+                if name.startswith("test_"):
+                    validate_legacy_dataset(matches[0], self.expected_test_samples)
+                elif name.startswith("train_"):
+                    validate_legacy_dataset(matches[0], self.expected_train_samples)
+                elif name.startswith("messages_"):
+                    validate_messages_dataset(matches[0], self.expected_train_samples)
+
+            node_p07_file = os.path.join("node_datasets_*", "knowledge_new_p07.jsonl")
+            node_p10_file = os.path.join("node_datasets_*", "knowledge_new_p10.jsonl")
+            for name in [
+                "skills_recipe_*.yaml",
+                "skills_train_*.jsonl",
+                "knowledge_recipe_*.yaml",
+                "knowledge_train_msgs_*.jsonl",
+                node_p07_file,
+                node_p10_file,
+            ]:
+                matches = glob.glob(os.path.join(self.tmp_path, name))
+                assert len(matches) == 1
+                if name.endswith("knowledge_new_p07.jsonl") or name.endswith(
+                    "knowledge_new_p10.jsonl"
+                ):
+                    validate_phase_leaf_node_dataset(matches[0])
+                elif name.startswith("skills_recipe_") or name.startswith(
+                    "knowledge_recipe_"
+                ):
+                    validate_recipe(matches[0])
+                elif name.startswith("skills_train_msgs_") or name.startswith(
+                    "knowledge_train_msgs_"
+                ):
+                    validate_mixed_dataset(matches[0])
+
+            for name in [
+                "knowledge_new_task.yaml",
+                "mmlubench_knowledge_new.jsonl",
+            ]:
+                matches = glob.glob(
+                    os.path.join(self.tmp_path, "node_datasets_*", name)
+                )
+                assert len(matches) == 1
+                if name == "knowledge_new_task.yaml":
+                    validate_lm_eval_task(matches[0])
+                elif name == "mmlubench_knowledge_new.jsonl":
+                    validate_mmlubench_dataset(matches[0])
```

Comment:
I don't believe the rest of this code needs to sit under the `with` statement, just the call to `generate data()`

Reply:
Valid point, shifted block left